python-package/lightgbm/basic.py

```diff
@@ -1262,8 +1482,15 @@ def _lazy_init(self, data, label=None, reference=None,
             self.__init_from_csc(data, params_str, ref_dataset)
         elif isinstance(data, np.ndarray):
             self.__init_from_np2d(data, params_str, ref_dataset)
-        elif isinstance(data, list) and len(data) > 0 and all(isinstance(x, np.ndarray) for x in data):
-            self.__init_from_list_np2d(data, params_str, ref_dataset)
+        elif isinstance(data, list) and len(data) > 0:
+            if all(isinstance(x, np.ndarray) for x in data):
+                self.__init_from_list_np2d(data, params_str, ref_dataset)
+            elif all(Sequence.is_class(x) for x in data):
```

Comment:
`isinstance()` returns `True` for subclasses. Is this what `is_class` was designed for?
```suggestion
            elif all(isinstance(x, Sequence) for x in data):
```

Reply:
This is to allow user to implement their own `Sequence` without subclassing. The only interface required now is `__len__` and `__getitem__`.

`Sequence` is an abstract class and uses `isinstance` here at first in this PR. We removed relying on `abc` and switched to this `Sequence.is_class` function later.

Reply:
Sorry, I'm not 100% sold on this.

For your example one can simply write something like the following:
```python
try:
    from lightgbm import Sequence as lgb_Sequence
except ImportError:
    class lgb_Sequence:
        pass

...


class MySequence(lgb_Sequence):
   ...
```
and a library will support LightGBM `Sequence` interface only if LightGBM is installed.

I strongly believe that first version "Sequence is an abstract class and uses `isinstance` here at first in this PR" was better design.

Reply:
I agree making `Sequence` an abstract class is better. The `Sequence.is_class` method is tricky to get right. I'll change `Sequence` back to an abstract class.